TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+        /// </summary>
+        internal CfgSectionSymbol CurrentSection
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current paragraph symbol.
+        /// </summary>
+        internal CfgParagraphSymbol CurrentParagraph
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The current Sentence in the current Paragraph.
+        /// </summary>
+        internal CfgSentence CurrentSentence
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The Section and Paragraph Domain of this program.
+        /// </summary>
+        internal Container<Symbol> SectionsParagraphs
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Ordered list of all Sections an Paragraphs encountered in order.
+        /// </summary>
+        internal List<Symbol> AllSectionsParagraphs
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// All pending Goto instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        protected LinkedList<Tuple<Goto, BasicBlockForNode>> PendingGOTOs;
+        /// <summary>
+        /// All encountered PERFORM procedure instructions
+        /// </summary>
+        protected LinkedList<Tuple<PerformProcedure, BasicBlockForNodeGroup>> PendingPERFORMProcedures;
+        /// <summary>
+        /// Pending ALTER instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        protected LinkedList<Tuple<Alter, BasicBlockForNode>> PendingALTERs;
+
+        /// <summary>
+        /// More Symbol Reference associated to altered GOTO statement.
+        /// </summary>
+        protected Dictionary<Goto, HashSet<SymbolReference>> PendingAlteredGOTOS;
+
+        /// <summary>
+        /// All pending Next Sentence instructions that will be handled at the end of the Procedure Division.
+        /// </summary>
+        internal LinkedList<Tuple<NextSentence, BasicBlockForNode, CfgSentence>> PendingNextSentences;
+
+        /// <summary>
+        /// All encountered sentences
+        /// </summary>
+        internal List<CfgSentence> AllSentences;
+
+        public IList<Diagnostic> Diagnostics { get; private set; }
+        /// <summary>
+        /// Empty constructor.
+        /// </summary>
+        public ControlFlowGraphBuilder() : this(null)
+        {
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="parentCfgBuilder">Parent Control Flow Builder for a nested program</param>
+        public ControlFlowGraphBuilder(ControlFlowGraphBuilder<D> parentCfgBuilder = null)
+        {
+            this.ParentProgramCfgBuilder = parentCfgBuilder;
+            this.Diagnostics = new List<Diagnostic>();
+            this.UseEvaluateCascade = true;
+            this.UseSearchCascade = true;
```

Comment:
`UseEvaluateCascade` and `UseSearchCascade` are always set to true. Should we keep these properties ? If so, they should be added as parameters in the constructor.

Reply:
The first implementation was using a CFG graph construction with basic block in sequence in sequence for WHENs and WHENOTHER alternatives, but the best way followed by IBM compiler when the code is generated is to build corresponding basic blocks in waterfall (Waterfall IF THEN ELSE construct). So these options are left there in order to be able to test both styles.